core-backend/src/runtime.rs

```diff
@@ -179,49 +179,52 @@ impl<'a, 'b, Ext: BackendExternalities + 'static> CallerWrap<'a, 'b, Ext> {
             .decrease_current_counter_to(gas_counter);
         res
     }
-}
 
-impl<'a, 'b, Ext> MemoryAccessRecorder for CallerWrap<'a, 'b, Ext> {
-    fn register_read(&mut self, ptr: u32, size: u32) -> WasmMemoryRead {
-        self.manager.register_read(ptr, size)
+    pub fn set_termination_reason(&mut self, reason: UndefinedTerminationReason) {
+        self.host_state_mut().termination_reason = reason;
     }
 
-    fn register_read_as<T: Sized>(&mut self, ptr: u32) -> WasmMemoryReadAs<T> {
-        self.manager.register_read_as(ptr)
-    }
-
-    fn register_read_decoded<T: Decode + MaxEncodedLen>(
+    /// Process fallible syscall function result
+    pub fn process_fallible_func_result<T: Sized>(
         &mut self,
-        ptr: u32,
-    ) -> WasmMemoryReadDecoded<T> {
-        self.manager.register_read_decoded(ptr)
-    }
-
-    fn register_write(&mut self, ptr: u32, size: u32) -> WasmMemoryWrite {
-        self.manager.register_write(ptr, size)
-    }
-
-    fn register_write_as<T: Sized>(&mut self, ptr: u32) -> WasmMemoryWriteAs<T> {
-        self.manager.register_write_as(ptr)
-    }
-}
-
-impl<Ext: BackendExternalities + 'static> BackendState for CallerWrap<'_, '_, Ext> {
-    fn set_termination_reason(&mut self, reason: UndefinedTerminationReason) {
-        self.host_state_mut().set_termination_reason(reason);
+        res: Result<T, RunFallibleError>,
+    ) -> Result<Result<T, u32>, UndefinedTerminationReason> {
+        match res {
+            Err(RunFallibleError::FallibleExt(ext_err)) => {
+                let code = ext_err.to_u32();
+                log::trace!(target: "syscalls", "fallible syscall error: {ext_err}");
+                Ok(Err(code))
+            }
+            Err(RunFallibleError::UndefinedTerminationReason(reason)) => Err(reason),
+            Ok(res) => Ok(Ok(res)),
+        }
+    }
+
+    /// Process alloc function result
+    pub fn process_alloc_func_result<T: Sized, ExtAllocError: BackendAllocSyscallError>(
+        &mut self,
+        res: Result<T, ExtAllocError>,
+    ) -> Result<Result<T, ExtAllocError>, UndefinedTerminationReason> {
+        match res {
+            Ok(t) => Ok(Ok(t)),
+            Err(err) => match err.into_backend_error() {
+                Ok(ext_err) => Err(ext_err.into()),
+                Err(alloc_err) => Ok(Err(alloc_err)),
+            },
+        }
     }
 }
 
-impl<'a, 'b, Ext: BackendExternalities + 'static> MemoryOwner for CallerWrap<'a, 'b, Ext> {
-    fn read(&mut self, read: WasmMemoryRead) -> Result<Vec<u8>, MemoryAccessError> {
+impl<'a, 'b, Ext: BackendExternalities + 'static> CallerWrap<'a, 'b, Ext> {
```

Comment:
if you don't want to mess up impls than at least add the comments whats being implemented below